File: 979.cpp
struct Solution {
    int dfs(TreeNode *node, int &total){
        if (node == nullptr){
            return 0;
        }
        int res = node->val - 1, subres = dfs(node->left, total);
        total += abs(subres);
        res += subres;
        subres = dfs(node->right, total);
        total += abs(subres);
        res += subres;
        return res;
    }

    inline int distributeCoins(TreeNode* root) {
        int res = 0;
        dfs(root, res);
        return res;
    }
};

File: sol/979/979.py
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        totalMoves = 0
        def residual(node):
            if node is None:
                return 0
            lres = residual(node.left)
            rres = residual(node.right)
            nonlocal totalMoves
            totalMoves += abs(lres) + abs(rres)
            return lres + rres + node.val - 1
        residual(root)
        return totalMoves
